### ISSUE TITLE
Add MySQL Reserved Word 'lead'

### ISSUE THREAD

**Issue by mreschke** [sqlalchemy/sqlalchemy#7167]
### Describe the bug

Using MariaDB 10.6.4 with a database table called `lead`.  Turns out this is a new reserved word in MariaDB.   I am using SQLAlchemy 1.3.24 and adding `lead` to your `dialects/mysql/base.py` `RESERVED_WORDS` set fixes the issue, causing the `lead` table to be escaped properly.  Looking at your latest 1.4.x branch, `lead` is not in there either.

I know 1.3.24 is old, but still used in the industry.  If you could backport that reserved word in 1.3 that would be appreciated.

### To Reproduce

```python
Run a JOIN on a table called `lead` on MariaDB 10.6.4.  An example SQL produced will be 

SELECT DISTINCT opportunities.id, opportunities.name, opportunities.lead_id, lead.id AS "lead__id", lead.name
AS "lead__name", lead.primary_contact_id AS "lead__primary_contact_id" FROM opportunities LEFT OUTER JOIN
leads AS lead ON opportunities.lead_id = lead.id;
```

Which fails with `"You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'lead ON opportunities.lead_id = lead.id'` because `lead` is not escaped properly as it is not in your set of RESERVED words.
```


### Error

```
"You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'lead ON opportunities.lead_id = lead.id'
``` 

### Versions

- OS: Linux Manjaro
- Python: 3.9.0
- SQLAlchemy: 1.3.24
- Database: MariaDB 10.6.4
- DBAPI (eg: psycopg, cx_oracle, mysqlclient): aiomysql via encode/databases


### Additional context

_No response_

**Comment by mreschke** [comment]
You can see all mariadb 10.6 reserved words here - https://github.com/MariaDB/server/blob/10.6/sql/lex.h

**Comment by mreschke** [comment]
While you're at it, might as well add all the other missing reserved words from https://github.com/MariaDB/server/blob/10.6/sql/lex.h, more than just `lead`.  All would cause issues for someone down the line and require proper escaping.  Thanks for the help.

**Comment by CaselIT** [comment]
If you can provide a pr it would be great! Thanks

**Comment by zzzeek** [comment]
hey there -

I'm not able to reproduce your failure on mariadb 10.6.4:

```
Welcome to the MariaDB monitor.  Commands end with ; or \g.
Your MariaDB connection id is 11226
Server version: 10.6.4-MariaDB MariaDB Server

Copyright (c) 2000, 2018, Oracle, MariaDB Corporation Ab and others.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

MariaDB [test]> create table leads (id integer, name varchar(20), primary_contact_id integer);
Query OK, 0 rows affected (0.028 sec)

MariaDB [test]> create table opportunities (id integer, name varchar(20), lead_id integer);
Query OK, 0 rows affected (0.023 sec)

MariaDB [test]> SELECT DISTINCT opportunities.id, opportunities.name, opportunities.lead_id, lead.id AS "lead__id", lead.name AS "lead__name", lead.primary_contact_id AS "lead__primary_contact_id" FROM opportunities LEFT OUTER JOIN leads AS lead ON opportunities.lead_id = lead.id;
Empty set (0.002 sec)
```

the addition of reserved words seems enormous and beside the above I'm not able to reproduce any problem related to a random sample of words,  "engines", "engine", "fast", and "errors".  I'm able to use these words as table / column names without quoting in mariadb 10.6 and there's no issue:

```
Your MariaDB connection id is 11224
Server version: 10.6.4-MariaDB MariaDB Server

Copyright (c) 2000, 2018, Oracle, MariaDB Corporation Ab and others.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

MariaDB [test]> create table engines (engine varchar(20), errors varchar(30), fast integer);
Query OK, 0 rows affected (0.025 sec)

MariaDB [test]> show create table engines;
+---------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table   | Create Table                                                                                                                                                             |
+---------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| engines | CREATE TABLE `engines` (
  `engine` varchar(20) DEFAULT NULL,
  `errors` varchar(30) DEFAULT NULL,
  `fast` int(11) DEFAULT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 |
+---------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.002 sec)

MariaDB [test]> insert into engines (engine, errors, fast) values ('x', 'y', 10);
Query OK, 1 row affected (0.001 sec)

MariaDB [test]> select engine, errors, fast from engines;
+--------+--------+------+
| engine | errors | fast |
+--------+--------+------+
| x      | y      |   10 |
+--------+--------+------+
1 row in set (0.001 sec)

MariaDB [test]> 
```

I think this issue with "lead" is specific to that name and how it's being used (of course it's very poor form for mariadb / mysql that it accepts a reserved word in any context).   I tried a similar idea with a few more reserved words "auto" and "committed", no issue:

```
MariaDB [test]> select * from engines as auto join engines as committed on auto.engine = committed.fast;
Empty set, 1 warning (0.002 sec)
```

can you show what your mariadb settings are namely SQL_MODE ?

**Comment by zzzeek** [comment]
Ok at this point, I would need a reproducer to understand the problem.  if there's an odd SQL_MODE setting in play or something, we could detect that instead.   right now i can't reproduce on the same version of the server so i dont want to move on this until we understand the problem.

**Comment by kkirsche** [comment]
https://mariadb.com/kb/en/reserved-words/
https://dev.mysql.com/doc/refman/8.0/en/keywords.html

are the two reserved words list I found (note the MySQL is version dependent) and differs from the lexer in critical ways.

**Comment by kkirsche** [comment]
> hey there -
> 
> I'm not able to reproduce your failure on mariadb 10.6.4:
> 
> ```
> Welcome to the MariaDB monitor.  Commands end with ; or \g.
> Your MariaDB connection id is 11226
> Server version: 10.6.4-MariaDB MariaDB Server
> 
> Copyright (c) 2000, 2018, Oracle, MariaDB Corporation Ab and others.
> 
> Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.
> 
> MariaDB [test]> create table leads (id integer, name varchar(20), primary_contact_id integer);
> Query OK, 0 rows affected (0.028 sec)
> 
> MariaDB [test]> create table opportunities (id integer, name varchar(20), lead_id integer);
> Query OK, 0 rows affected (0.023 sec)
> 
> MariaDB [test]> SELECT DISTINCT opportunities.id, opportunities.name, opportunities.lead_id, lead.id AS "lead__id", lead.name AS "lead__name", lead.primary_contact_id AS "lead__primary_contact_id" FROM opportunities LEFT OUTER JOIN leads AS lead ON opportunities.lead_id = lead.id;
> Empty set (0.002 sec)
> ```
> 
> the addition of reserved words seems enormous and beside the above I'm not able to reproduce any problem related to a random sample of words, "engines", "engine", "fast", and "errors". I'm able to use these words as table / column names without quoting in mariadb 10.6 and there's no issue:
> 
> ```
> Your MariaDB connection id is 11224
> Server version: 10.6.4-MariaDB MariaDB Server
> 
> Copyright (c) 2000, 2018, Oracle, MariaDB Corporation Ab and others.
> 
> Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.
> 
> MariaDB [test]> create table engines (engine varchar(20), errors varchar(30), fast integer);
> Query OK, 0 rows affected (0.025 sec)
> 
> MariaDB [test]> show create table engines;
> +---------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
> | Table   | Create Table                                                                                                                                                             |
> +---------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
> | engines | CREATE TABLE `engines` (
>   `engine` varchar(20) DEFAULT NULL,
>   `errors` varchar(30) DEFAULT NULL,
>   `fast` int(11) DEFAULT NULL
> ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 |
> +---------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
> 1 row in set (0.002 sec)
> 
> MariaDB [test]> insert into engines (engine, errors, fast) values ('x', 'y', 10);
> Query OK, 1 row affected (0.001 sec)
> 
> MariaDB [test]> select engine, errors, fast from engines;
> +--------+--------+------+
> | engine | errors | fast |
> +--------+--------+------+
> | x      | y      |   10 |
> +--------+--------+------+
> 1 row in set (0.001 sec)
> 
> MariaDB [test]> 
> ```
> 
> can you show what your mariadb settings are namely SQL_MODE ?

Per https://mariadb.com/kb/en/reserved-words/ `leads` and `lead` is not a keyword, `leading` is though.

**Comment by kkirsche** [comment]
I think you mixed up who was commenting @zzzeek

**Comment by CaselIT** [comment]
I initially suggested a separate file since the initial list in the closed pr was very long, like 500 more lines.

The current version seems more manageable, so I guess it's ok to keep it in the same file

**Comment by zzzeek** [comment]
> I initially suggested a separate file since the initial list in the closed pr was very long, like 500 more lines.
> 
> The current version seems more manageable, so I guess it's ok to keep it in the same file

yes from my POV if our reserved words lists are heading into the many hundreds, we'd need to think of other approaches towards reserved words.   in any case this does not seem to really be a "reserved word" list because I'm able to use these words in SQL statements without any quoting, meaning those words are not "reserved" in the usual sense.

**Comment by zzzeek** [comment]
OK great, it is exactly the word "lead" on MySQL 8.  

so now we get a grasp on really good reserved word lists for both DBs.  MariaDB's is at: https://mariadb.com/kb/en/reserved-words/ , MySQL is at https://dev.mysql.com/doc/refman/8.0/en/keywords.html , those words marked with "(R)" or "reserved".  

so for @kkirsche what I would like most would be a reserved word approach in the MySQL dialect that is selective based on whether or not the dialect has the "is_mariadb" flag set.    since this is a special case it's not necessarily unreasonable these two lists would be in a separate file.

**Comment by kkirsche** [comment]
Sounds good @zzzeek, I should be able to implement that tomorrow.

If I understand correctly, my plan will be to:

1. Move the `RESERVED_WORDS` to a separate file, `reserved_words.py` from `base.py`
2. Split the list of reserved words into individual `set`'s based on whether it's from MariaDB or MySQL.
3. Populate the lists using the linked URLs (and lower versions as applicable for MySQL), and include comments for versions. 
4. Include links to the reserved word lists used to populate the list in a comment on the header.
5. For version specific introductions for the MySQL list, note in comments from what version introduced the keyword or when it was removed, if applicable. 
6. Set the list of reserved words conditionally based on the value of `is_mariadb` (will need to look where that is but I assume it's on base or easily locatable)

This implementation will fully replace the existing list. I will break this up, commit wise, such that moving and the lists are separate actions.

Please let me know if I misunderstood anything.

**Comment by kkirsche** [comment]
Using code above as foundation, beginning update of the pull request, and then will manually verify the automated list generation. Please let me know if you'd like me to put the automation anywhere, once validated if confirmed it works in full

**Comment by kkirsche** [comment]
Maybe not, but it's still much much faster than trying to write the list out from the upstream by hand :)

**Comment by kkirsche** [comment]
Long story short, I'm done performing work until reviewed.

**Comment by CaselIT** [comment]
> I would consider automation of reserved word lists and whatnot to be wholly separate issues.

agree. will run it in gerrit to see it the pipeline passes there (testing against mysql db)

**Comment by iangilfillan** [comment]
Thanks for reporting https://jira.mariadb.org/browse/MDEV-26916. The cause here is that SYSTEM is not actually a MariaDB reserved word, but in the context of ALTER TABLE ADD SYSTEM ... the parser expects this to be followed by VERSIONING, as per https://mariadb.com/kb/en/system-versioned-tables/. Perhaps it should be added as a MariaDB reserved word, but SQLAlchemy should probably add it to the list to avoid possible issues.

**Comment by kkirsche** [comment]
Thank you, Ian, for the feedback about the use of SYSTEM as a reserved word.

I know this is a complex issue on the MariaDB side, especially as there are different ways to define what a reserved word means, so we appreciate your, Sergei, and the rest of the team's time looking into this
